### PR TITLE
Fix CLS propagation for all Bluebird methods

### PR DESCRIPTION
**Pull request by mdarveau**
This is a fix for https://github.com/sequelize/sequelize/issues/3686.

I decided to shim only the documented bluebird API so it's less likely to break due to internal changes.
I basically shimmed all bluebird methods that accepts functions except `try` and `method` since the former call the function synchronously and the later just wrap the function.

I tested the change with my code base and it works. To be honest, I most probably have no idea what I'm doing here so a thorough review is more than welcome.

### CODE REVIEW

File: lib/promise.js [hunk 1]
```diff
@@ -1,17 +1,44 @@
...
+
+      return fn.apply(this, arguments);
+    };
+  });
+}
+
+// Core
+shimCLS(Promise, 'join', [-1]);
+shimCLS(Promise.prototype, 'then', [0, 1, 2]);
+shimCLS(Promise.prototype, 'spread', [0, 1]);
```
Comment (mickhansen) [inline lib/promise.js]: Not sure you need to shim `spread/catch/error/finally`.
Might want to add some tests so we can be sure though.

Comment (janmeier) [inline lib/promise.js]: Shouldn't be needed, I think they all map to `then` internally

**Comment by mdarveau** [comment]
@mickhansen I was afraid you would say that :-) Will add some tests tonight.

@janmeier "Shouldn't be needed, I think they all map to then internally". Is this comment for the whole pull request or just some shim? Note that I no longer shim the internal `_then` but the public `then`. I went the way of shimming only the published API entry points so we don't rely on any bluebird internal implementation anymore.

**Comment by mdarveau** [comment]
@mickhansen I added tests for all shimmed methods. I can now say that I called most bluebird methods :-)

@janmeier You are right, only the collection methods needed to be shimmed. Here is the result of the tests with the original code:
```
  [POSTGRES] Continuation local storage
    bluebird shims
      ✓ join
      ✓ then
      ✓ spread
      ✓ catch
      ✓ error
      ✓ finally
      1) map
      2) reduce
      3) filter
      4) each
      ✓ tap
```

However, I think it's more robust to shim the public methods instead of the `_then` method. If you do not want to go that way, the minimum shim would be:
```
// Core
shimCLS(Promise.prototype, '_then', [0, 1, 2]);

// Collections
shimCLS(Promise.prototype, 'map', [0]);
shimCLS(Promise.prototype, 'reduce', [0]);
shimCLS(Promise.prototype, 'filter', [0]);
shimCLS(Promise.prototype, 'each', [0]);
```

I could not find any common point for all collections methods. However, I'm a but lost in Bluebird code and I'm not saying it doesn't exist. But again, what's the point of binding ourself to internal implementation when it can be avoided.

**Comment by mickhansen** [comment]
Theres two levels to robustness i suppose, with `_then` we risk `_then` changing but any new methods adding using it automaticall works.

If we shim each method seperately we don't shim an internal but we have N amount of public interfcaces that might change.

The bluebird code is super optimized and a lot of small methods so can be hard to follow :)

**Comment by janmeier** [comment]
After giving it some more thought, I don't think it matters much which way we choose - shimming the public API is fine by me.

File: test/integration/cls.test.js [hunk 2]
```diff
@@ -138,5 +138,150 @@ if (current.dialect.supports.transactions) {
...
+      });
+
+      it('then', function () {
+        var self = this;
+        return this.sequelize.transaction(function () {
+          var tid = self.ns.get('transaction').id;
+          return self.User.findAll().then(function () {
+            expect(self.ns.get('transaction').id).to.be.ok;
+            expect(self.ns.get('transaction').id).to.equal(tid);
+          },function () {
```
Comment (janmeier) [inline test/integration/cls.test.js]: Why provide 3 callbacks here - only one of those will ever be called anyway :)

**Comment by mickhansen** [comment]
Shimming the public API is fine as long as we remember to then add tests for each thing we shim.

**Comment by mdarveau** [comment]
I splitted the tests for fulfilled/rejected on then.

**Comment by mdarveau** [comment]
Do you think this will make it to 3.0.0?

**Comment by mickhansen** [comment]
Well right now it's very much red and failing (jshint errors).
But yeah likely if it's merge ready by tommorrow.
Else we'll just put it out with 3.0.1

**Comment by mdarveau** [comment]
All green now. Thanks!

BTW, if I go to https://github.com/sequelize/sequelize/releases, the latest release is `2.0.4` but I see tags up to `2.1.4`. Is this expected?

**Comment by mickhansen** [comment]
@mdarveau no idea how those releases are generated actually ;p

File: test/integration/cls.test.js [hunk 3]
```diff
@@ -138,5 +138,155 @@ if (current.dialect.supports.transactions) {
         });
       });
     });
+
+    describe.only('bluebird shims', function () {
```
Comment (mickhansen) [inline test/integration/cls.test.js]: Remove `.only`

**Comment by mdarveau** [comment]
Oh so sorry. The next PR will (hopefully) be easier.

Fixed now, thanks.

**Comment by mickhansen** [comment]
Alright, looking good @mdarveau - Now if i can just get you to squash this to a more descriptive set of (or single) commits we can get it merged :)

**Comment by mdarveau** [comment]
Done. The build is failing, but just like the master.

Thanks!

**Comment by mickhansen** [comment]
Thank you @mdarveau
